src/conn_state.c

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright 2019 Xaptum, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "conn_state.h"
+
+#include <errno.h>
+#include <fcntl.h>
+#include <string.h>
+#include <unistd.h>
+
+#include <arpa/inet.h>
+#include <net/if.h>
+#include <netinet/in.h>
+#include <sys/socket.h>
+
+#include "context.h"
+#include "log.h"
+#include "sockaddr.h"
+#include "udp.h"
+
+static
+int
+check_preferred_route(struct enftun_conn_state* conn_state)
+{
+    struct enftun_context* ctx = (struct enftun_context*)conn_state->data;
+    int rc = enftun_udp_connect_addr(&conn_state->udp, &ctx->tls.sock.remote_addr);
+    if (0 != rc)
+        return rc;
+
+    enftun_udp_close(&conn_state->udp);
+
+    rc = enftun_sockaddr_equal(&conn_state->udp.local_addr, &ctx->tls.sock.local_addr);
+    if (0 != rc)
+        conn_state->reconnect_cb(conn_state);
+
+    return rc;
+}
+
+static
+void
+on_poll(uv_poll_t* handle, int status, int events)
+{
+    (void) events;
+    struct enftun_conn_state* conn_state = handle->data;
+
+    char msg_buf[8192];
+
+    if (status < 0)
+        return;
+    if (0 == status)
```

Comment:
Can `status` be greater than `0`?

If so, what should we do?

If not (or we don't care) then remove the `if (0 == status)` check and always do the read_message/check_preferred stuff.